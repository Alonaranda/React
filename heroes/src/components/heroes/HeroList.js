import React, {useMemo} from 'react';
import { getHeroesByPublisher } from '../../selectors/getHeroesByPublisher';
import { HeroeCard } from './HeroeCard';

export const HeroList = ({publisher}) => {
    const heroes = useMemo(() => getHeroesByPublisher(publisher), [publisher])
    //const heroes = getHeroesByPublisher(publisher);



    return (
        <div className="card-columns animate__fadeInDownBig">
            {
                heroes.map(hero => (
                    <HeroeCard 
                        key={hero.id}
                        {...hero}
                    />
                ))
            }
        </div>
    )
}
