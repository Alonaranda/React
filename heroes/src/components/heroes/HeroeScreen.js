import React, {Â useMemo } from 'react';
import {useParams, Redirect} from 'react-router-dom';
import { getHeroesById } from '../../selectors/getHeroesById';

export const HeroeScreen = ({history}) => {

    const {heroeId} = useParams();
    
    const hero = useMemo(() => getHeroesById(heroeId), [heroeId])
    //const hero = getHeroesById(heroeId);
    //console.log(hero);

    if(!hero){
        return <Redirect to="/"/>
    }

    const {
        superhero,
        publisher,
        alter_ego,
        first_appearance,
        characters
    } = hero;

    const handleReturn = () =>{
        if(history.length <=2){
            history.push('/');
        }else{
            history.goBack();
        }
    }

    return (
        <div className="row mt-5 animate__fadeInTopLeft">
            <div className="col-4">
                <img 
                    className="img-thumbnail animate__fadeInDownBig" 
                        alt={superhero} 
                        src={`../assets/heroes/${heroeId}.jpg`}
                    />
            </div>
            <div className="col-8">
                <h3 className="text-success">{superhero}</h3>
                <ul className="list-group list-group-flush"> 
                    <li className="list-group-item"><b>Alter ego:</b> {alter_ego}</li>
                    <li className="list-group-item"><b>Publisher:</b> {publisher}</li>
                    <li className="list-group-item"><b>First appearance:</b> {first_appearance}</li>
                </ul>
                <h5 className="mt-3"> Character:</h5>
                <p>{characters}</p>
                <button 
                    className="btn btn-outline-info mt-2"
                    onClick={handleReturn}
                >Return</button>
            </div>
        </div>
    )
}
