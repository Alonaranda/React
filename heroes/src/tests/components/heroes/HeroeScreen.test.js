import React from 'react';
import { mount } from 'enzyme';
import { HeroeScreen } from '../../../components/heroes/HeroeScreen';
import { MemoryRouter, Route } from 'react-router-dom';

describe('Probando HeroeScreen', () => {

    const history = {
        length: 10,
        goBack: jest.fn(),
        push: jest.fn()

    }
    
    test('Debe de mostrar el componente redirect ', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero']}>
                <HeroeScreen history={history}/>
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('Redirect').exists()).toBe(true);
    });

    test('Debe de mostrar un heroe si el parametro existe', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe/marvel-hulk']}>
                <Route path="/heroe/:heroeId" component={HeroeScreen}/>
            </MemoryRouter>
        );

        expect(wrapper.find('.row').exists()).toBe(true);
    });

    test('Debe regresar a la pantalla anterior con PUSH', () => {
        const history = {
            length: 1,
            goBack: jest.fn(),
            push: jest.fn()
    
        }
        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe/marvel-hulk']}>
                <Route 
                    path="/heroe/:heroeId" 
                    component={() => <HeroeScreen history={history}/>}
                />
            </MemoryRouter>
        );
        wrapper.find('button').prop('onClick')();
        expect(history.push).toHaveBeenCalledWith('/');
        expect(history.goBack).not.toHaveBeenCalled();
    });

    test('Debe regresar a la pantalla anterior con goBack', () => {
        const history = {
            length: 3,
            goBack: jest.fn(),
            push: jest.fn()
    
        }
        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe/marvel-hulk']}>
                <Route 
                    path="/heroe/:heroeId" 
                    component={() => <HeroeScreen history={history}/>}
                />
            </MemoryRouter>
        );
        wrapper.find('button').prop('onClick')();
        expect(history.goBack).toHaveBeenCalledWith();
        expect(history.push).not.toHaveBeenCalled();
    });

    test('Debe de llamar el Redirect', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe/marvel-huwdefek']}>
                <Route 
                    path="/heroe/:heroeId" 
                    component={() => <HeroeScreen history={history}/>}
                />
            </MemoryRouter>
        );
        console.log(wrapper.html());
        expect(wrapper.find('Redirect').exists()).toBe(false);
    });
});