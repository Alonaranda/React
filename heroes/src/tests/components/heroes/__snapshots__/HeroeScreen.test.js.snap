// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando HeroeScreen Debe de mostrar el componente redirect  1`] = `""`;
